beater/tracing.go

```diff
@@ -19,112 +19,30 @@ package beater
-// to send tracing data to itself, it will return a tracerServer
-// that can be used for receiving trace data.
-func initTracer(info beat.Info, cfg *config.Config, logger *logp.Logger) (*apm.Tracer, *tracerServer, error) {
-	if !cfg.SelfInstrumentation.IsEnabled() {
-		os.Setenv("ELASTIC_APM_ACTIVE", "false")
-		logger.Infof("self instrumentation is disabled")
-	} else {
-		os.Setenv("ELASTIC_APM_ACTIVE", "true")
-		logger.Infof("self instrumentation is enabled")
-	}
-	if !cfg.SelfInstrumentation.IsEnabled() {
-		tracer, err := apm.NewTracer(info.Beat, info.Version)
-		return tracer, nil, err
-	}
-
-	if cfg.SelfInstrumentation.Profiling.CPU.IsEnabled() {
-		interval := cfg.SelfInstrumentation.Profiling.CPU.Interval
-		duration := cfg.SelfInstrumentation.Profiling.CPU.Duration
-		logger.Infof("CPU profiling: every %s for %s", interval, duration)
-		os.Setenv("ELASTIC_APM_CPU_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
-		os.Setenv("ELASTIC_APM_CPU_PROFILE_DURATION", fmt.Sprintf("%dms", int(duration.Seconds()*1000)))
-	}
-	if cfg.SelfInstrumentation.Profiling.Heap.IsEnabled() {
-		interval := cfg.SelfInstrumentation.Profiling.Heap.Interval
-		logger.Infof("Heap profiling: every %s", interval)
-		os.Setenv("ELASTIC_APM_HEAP_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
-	}
-
-	var tracerTransport transport.Transport
-	var tracerServer *tracerServer
-	if cfg.SelfInstrumentation.Hosts != nil {
-		// tracing destined for external host
-		t, err := transport.NewHTTPTransport()
-		if err != nil {
-			return nil, nil, err
-		}
-		t.SetServerURL(cfg.SelfInstrumentation.Hosts...)
-		if cfg.SelfInstrumentation.APIKey != "" {
-			t.SetAPIKey(cfg.SelfInstrumentation.APIKey)
-		} else {
-			t.SetSecretToken(cfg.SelfInstrumentation.SecretToken)
-		}
-		tracerTransport = t
-		logger.Infof("self instrumentation directed to %s", cfg.SelfInstrumentation.Hosts)
-	} else {
-		var err error
-		tracerServer, err = newTracerServer(cfg)
-		if err != nil {
-			return nil, nil, err
-		}
-		tracerTransport = tracerServer.transport
-	}
-
-	var environment string
-	if cfg.SelfInstrumentation.Environment != nil {
-		environment = *cfg.SelfInstrumentation.Environment
-	}
-	tracer, err := apm.NewTracerOptions(apm.TracerOptions{
-		ServiceName:        info.Beat,
-		ServiceVersion:     info.Version,
-		ServiceEnvironment: environment,
-		Transport:          tracerTransport,
-	})
-	if err != nil {
-		return nil, nil, err
-	}
-	tracer.SetLogger(logp.NewLogger(logs.Tracing))
-
-	return tracer, tracerServer, nil
-}
-
 type tracerServer struct {
-	cfg       *config.Config
-	logger    *logp.Logger
-	server    *http.Server
-	listener  net.Listener
-	transport transport.Transport
+	cfg      *config.Config
+	logger   *logp.Logger
+	server   *http.Server
+	listener net.Listener
 }
 
-func newTracerServer(cfg *config.Config) (*tracerServer, error) {
+func newTracerServer(cfg *config.Config, listener net.Listener) (*tracerServer, error) {
```

Comment:
```suggestion
func newTracerServer(cfg *config.Config, listener net.Listener) (*tracerServer) {
```

(AFAICS this function can't return any errors any more - right?)